packages/playground-editor/src/App.tsx

```diff
@@ -0,0 +1,139 @@
+import * as monaco from "monaco-editor";
+import { languages } from "monaco-editor";
+import types from "./types.json";
+import ScriptTarget = languages.typescript.ScriptTarget;
+import * as q from "groqd";
+import debounce from "lodash.debounce";
+import { emitError, emitQuery } from "./messaging";
+
+export function App() {
+  const containerRef = React.useRef<HTMLDivElement>(null);
+  const editorRef = React.useRef<monaco.editor.IStandaloneCodeEditor>();
+
+  /**
+   * Run the editor code to extract query, and then emit that out
+   */
+  const runCode = React.useCallback(async () => {
+    try {
+      const editor = editorRef.current;
+      if (!editor) throw new Error("Editor not yet instantiated");
+
+      const model = editor.getModel();
+      if (!model) throw new Error();
+
+      const worker = await monaco.languages.typescript.getTypeScriptWorker();
+      const client = await worker(model.uri);
+      const emitResult = await client.getEmitOutput(model.uri.toString());
+      const code = emitResult.outputFiles[0].text;
+
+      let result = "";
+      const libs = {
+        groqd: q,
+        playground: {
+          runQuery: (query: { query: string }) => {
+            try {
+              result = query.query;
+            } catch {}
+          },
+        },
+      };
+      const scope = {
+        exports: {},
+        require: (name: keyof typeof libs) => libs[name],
+      };
+      const keys = Object.keys(scope);
+      new Function(...keys, code)(
+        ...keys.map((key) => scope[key as keyof typeof scope])
+      );
+
+      // Emit the result
+      emitQuery(result);
+    } catch (err) {
+      console.error(err);
+      emitError(err instanceof Error ? err.message : "");
+    }
+  }, []);
+
+  const handleContentChange = React.useMemo(
+    () => debounce(runCode, 300),
+    [runCode]
+  );
+
+  React.useEffect(() => {
+    const container = containerRef.current;
+    if (!container || editorRef.current) return;
+
+    monaco.languages.typescript.typescriptDefaults.setCompilerOptions({
+      typeRoots: ["groqd", "zod"],
+      target: ScriptTarget.ES5,
+      strict: true,
+      skipLibCheck: true,
+      esModuleInterop: true,
+    });
+
+    const model = monaco.editor.createModel(
+      INIT_VALUE,
+      "typescript",
+      monaco.Uri.parse("file:///main.ts")
+    );
+
+    const didChangeInstance = model.onDidChangeContent(handleContentChange);
+
+    editorRef.current = monaco.editor.create(container, {
+      model,
+      language: "typescript",
+      minimap: { enabled: false },
+    });
+
+    monaco.languages.typescript.typescriptDefaults.setExtraLibs(extraLibs);
+
+    // Run the code on mount
+    runCode().catch(console.error);
+
+    return () => {
+      didChangeInstance.dispose();
+    };
+  }, []);
+
+  return (
+    <div ref={containerRef} style={{ width: "100vw", height: "100vh" }}></div>
+  );
+}
+
+// Initial code, will likely change in the future.
+const INIT_VALUE = [
```

Comment:
I think this is fine for now, but I wonder if consuming it from a template file might be better and easier to update?

Reply:
yeah agreed. At some point, we'll also allow passing in initial value via query param – so this is going to change in the near future.